src/components/cylc/tree/tree.js

```diff
@@ -0,0 +1,497 @@
+  }
+
+  clear () {
+    this.lookup.clear()
+    this.root = {
+      id: '',
+      node: {},
+      children: []
+    }
+  }
+
+  isEmpty () {
+    return this.lookup.size === 0
+  }
+
+  /**
+   * @param {null|{
+   *   id: string,
+   *   node: Object,
+   *   children: []
+   * }} node
+   */
+  recursivelyRemoveNode (node) {
+    if (node) {
+      const stack = [node]
+      while (stack.length > 0) {
+        const n = stack.pop()
+        this.lookup.delete(n.id)
+        if (n.children && n.children.length > 0) {
+          stack.push(...n.children)
+        }
+      }
+    }
+  }
+
+  // --- Cycle points
+
+  /**
+   * @param {null|{
+   *   id: string,
+   *   node: Object,
+   *   children: []
+   * }} cyclePoint
+   */
+  addCyclePoint (cyclePoint) {
+    if (cyclePoint && !this.lookup.has(cyclePoint.id)) {
+      this.lookup.set(cyclePoint.id, cyclePoint)
+      this.root.children.push(cyclePoint)
+      // sort cycle points
+      this.root.children.sort((cyclepoint, anotherCyclepoint) => {
+        return cyclepoint.id.localeCompare(anotherCyclepoint.id)
+      })
+    }
+  }
+
+  /**
+   * @param {null|{
+   *   id: string,
+   *   node: Object,
+   *   children: []
+   * }} cyclePoint
+   */
+  updateCyclePoint (cyclePoint) {
+    if (cyclePoint) {
+      const node = this.lookup.get(cyclePoint.id)
+      if (node) {
+        merge(node, cyclePoint)
+      }
+    }
+  }
+
+  /**
+   * @param {null|string} cyclePointId
+   */
+  removeCyclePoint (cyclePointId) {
+    const node = this.lookup.get(cyclePointId)
+    if (node) {
+      this.recursivelyRemoveNode(node)
+      this.root.children.splice(this.root.children.indexOf(node), 1)
+    }
+  }
+
+  tallyCyclePointStates () {
+    // calculate cycle point states
+    computeCyclePointsStates(this.root.children)
+  }
+
+  // --- Family proxies
+
+  /**
+   * @param {null|{
+   *   id: string,
+   *   node: Object,
+   *   children: []
+   * }} familyProxy
+   */
+  addFamilyProxy (familyProxy) {
+    // When we receive the families from the GraphQL endpoint, we are sorting by their
+    // firstParent's. However, you may get family proxies out of order when iterating
+    // them. When that happens, you may add a family proxy to the lookup, and only
+    // append it to the parent later.
+    if (familyProxy) {
+      // ignore the root family
+      if (familyProxy.id.endsWith(`|${FAMILY_ROOT}`)) {
+        return
+      }
+      // add if not in the lookup already
+      const existingFamilyProxy = this.lookup.get(familyProxy.id)
+      if (!existingFamilyProxy) {
+        this.lookup.set(familyProxy.id, familyProxy)
+      } else {
+        // We may get a family proxy added twice. The first time is when it is the parent of another
+        // family proxy. In that case, we create an orphan node in the lookup table.
+        // The second time will be node with more information, such as .firstParent {}. When this happens,
+        // we must remember to merge the objects.
```

Comment:
Ooh, nasty. Must remember to make sure views are resilient to this.

Reply:
I tried to add as many comments as possible, for each situation where I got an error like "Cannot use undefined...", or "blabla is not a function", etc... this kind of runtime unexpected errors.

For each error I either added a comment, or took it back to @dwsutherland to review if the UIS was giving the right data (hence the amount of re-work/review/answering-silly-questions that David had to do some weeks ago :sweat_smile: )